### PR TITLE
Allow user to set priority for a task

### PR DESCRIPTION
**Pull request by xiaoral2**
[(task:18336)](https://staging.teamyapp.com/teams/107/tasks/18336)
image
image
image
image
image

In the graph is the evidence simple compare before and after for sorting tasks with priority

### CODE REVIEW

File: src/client/entity/remoteUpdateTaskInput.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+export interface RemoteUpdateTaskInput {
+    goal: string;
+    context?: string;
+    ownerUserId?: string;
+    owningTeamId: string;
```
Comment (duanxianbo) [inline src/client/entity/remoteUpdateTaskInput.ts]: can we put the required field first and then put optional fields

File: src/components/internal/TaskListSections/UpcomingSection.tsx [hunk 2]
```diff
@@ -76,9 +76,13 @@ export class UpcomingSection extends Component<Props, State> {
     }
 
     public render() {
-        const flattenTasks = flattenGroupedTasksByOwner(
-            groupByOwners(this.props.tasks),
-        );
+        // const flattenTasks = flattenGroupedTasksByOwner(
+        //     groupByOwners(this.props.tasks),
+        // );
```
Comment (duanxianbo) [inline src/components/internal/TaskListSections/UpcomingSection.tsx]:
```suggestion
```

File: src/components/pages/Sprints.component.tsx [hunk 3]
```diff
@@ -103,6 +104,7 @@ export class SprintsComponent extends Component<Props, State> {
             sprintParticipants: [],
             sortOrders: {
                 upcoming: taskUrgencyOrder,
+                // upcoming: taskPriorityOrder,
```
Comment (duanxianbo) [inline src/components/pages/Sprints.component.tsx]:
```suggestion
```

**Review by duanxianbo — DISMISSED** [review]

**Comment by naturetime10** [comment]
Please put descriptions and screenshots in PR description instead of using multiple comments.

File: src/client/entity/input.ts [hunk 4]
```diff
@@ -39,6 +41,7 @@ export interface RemoteUpdateTaskInput {
     ownerUserId?: string;
     owningTeamId: string;
     effort?: string;
+    priority?: Priority;
```
Comment (naturetime10) [inline src/client/entity/input.ts:44]:
```suggestion
    priority?: string;
```

File: src/client/entity/remoteUpdateTaskInput.ts [hunk 5]
```diff
@@ -0,0 +1,9 @@
+export interface RemoteUpdateTaskInput {
```
Comment (naturetime10) [inline src/client/entity/remoteUpdateTaskInput.ts]: Let's use `RemoteUpdateTaskInput` from `input.ts` and remove this file.

File: .env.staging [hunk 6]
```diff
@@ -14,6 +14,8 @@ VITE_LONG_BACK_OFF_MAX_DELAY_MS=60000 # 1 minute
 VITE_IDENTITY_API_BASE_ENDPOINT=https://cloud.staging.teamyapp.com/identity
 VITE_IDENTITY_COOKIE_DOMAIN=staging.teamyapp.com
 VITE_IDENTITY_COOKIE_SECURE=true
+#VITE_IDENTITY_COOKIE_DOMAIN=localhost
+#VITE_IDENTITY_COOKIE_SECURE=false
```
Comment (naturetime10) [inline .env.staging]:
```suggestion
```

File: src/client/task.client.ts [hunk 7]
```diff
@@ -1,10 +1,8 @@
 import { GraphQLClient } from '@lib/network/GraphQL.client';
 
-import {
-    RemoteCreateTaskInput,
-    RemoteUpdateTaskInput,
-} from '@core/client/entity/input';
+import { RemoteCreateTaskInput } from '@core/client/entity/input';
```
Comment (naturetime10) [inline src/client/task.client.ts]:
```suggestion
import { RemoteCreateTaskInput, RemoteUpdateTaskInput } from '@core/client/entity/input';
```

File: src/client/task.client.ts [hunk 8]
```diff
@@ -1,10 +1,8 @@
...
 import { RemoteTask } from '@core/client/entity/remoteTask';
+import { RemoteUpdateTaskInput } from '@core/client/entity/remoteUpdateTaskInput';
```
Comment (naturetime10) [inline src/client/task.client.ts]:
```suggestion
```

File: src/components/internal/InlineTask.tsx [hunk 9]
```diff
@@ -1,4 +1,5 @@
 import classNames from 'classnames';
+import { unset } from 'lodash-es';
```
Comment (naturetime10) [inline src/components/internal/InlineTask.tsx]:
```suggestion
```

File: src/components/internal/TaskListSections/UpcomingSection.tsx [hunk 10]
```diff
@@ -76,9 +76,13 @@ export class UpcomingSection extends Component<Props, State> {
...
+        const sortOrder = this.getSortOrder();
+
```
Comment (naturetime10) [inline src/components/internal/TaskListSections/UpcomingSection.tsx]:
```suggestion
```

File: src/components/internal/TaskListSections/UpcomingSection.tsx [hunk 11]
```diff
@@ -76,9 +76,13 @@ export class UpcomingSection extends Component<Props, State> {
...
+        const sortedTasks = this.props.tasks.sort(sortOrder);
+
```
Comment (naturetime10) [inline src/components/internal/TaskListSections/UpcomingSection.tsx]:
```suggestion
```

File: src/components/internal/TaskListSections/UpcomingSection.tsx (hunk 2 above)
Comment (naturetime10) [inline src/components/internal/TaskListSections/UpcomingSection.tsx]: This is not need anymore.

File: src/data/task.order.ts [hunk 12]
```diff
@@ -1,14 +1,42 @@
...
+};
+
+const priorityOrder: Record<Priority, number> = {
+    URGENT: 1,
+    HIGH: 2,
+    MEDIUM: 3,
+    LOW: 4,
+};
+
 export const orderByPriority: SortOrder<Task> = (
```
Comment (naturetime10) [inline src/data/task.order.ts]:
```suggestion
export const orderByPriorityDesc: SortOrder<Task> = (
```

File: src/data/task.order.ts [hunk 13]
```diff
@@ -1,14 +1,42 @@
...
+    AWAITING: 4,
+    IN_PROGRESS: 5,
+    DELIVERED: 6,
+};
+
+const priorityOrder: Record<Priority, number> = {
+    URGENT: 1,
+    HIGH: 2,
+    MEDIUM: 3,
+    LOW: 4,
```
Comment (naturetime10) [inline src/data/task.order.ts]:
```suggestion
    LOW: 1,
    MEDIUM: 2,
    HIGH: 3,
    URGENT: 4,
```

File: src/data/task.order.ts [hunk 14]
```diff
@@ -1,14 +1,42 @@
...
+
+    if (!task1.priority) {
+        return 1;
+    }
+
+    if (!task2.priority) {
+        return -1;
+    }
+
+    return priorityOrder[task1.priority] - priorityOrder[task2.priority];
```
Comment (naturetime10) [inline src/data/task.order.ts]:
```suggestion
    return  priorityOrder[task2.priority] - priorityOrder[task1.priority];
```

File: src/data/task.order.ts [hunk 15]
```diff
@@ -1,14 +1,42 @@
...
+        return 0;
+    }
+
+    if (!task1.priority) {
+        return 1;
+    }
+
+    if (!task2.priority) {
+        return -1;
+    }
```
Comment (naturetime10) [inline src/data/task.order.ts:37]: Can you verify the ordering to make sure task with no priority with be sorted below the task which has priority?

File: src/entity/priority.ts [hunk 16]
```diff
@@ -0,0 +1 @@
+export type Priority = 'URGENT' | 'HIGH' | 'MEDIUM' | 'LOW';
```
Comment (naturetime10) [inline src/entity/priority.ts:1]:
```suggestion
export type Priority = 
    | 'URGENT' 
    | 'HIGH' 
    | 'MEDIUM' 
    | 'LOW';
```

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: src/client/entity/remoteUpdateTaskInput.ts (hunk 1 above)
Comment (xiaoral2) [reply src/client/entity/remoteUpdateTaskInput.ts]: Removed this file since we don't need it

File: src/client/entity/input.ts (hunk 4 above)
Comment (xiaoral2) [reply src/client/entity/input.ts:44]: solved in remoteTask.ts at line 23

File: src/components/pages/Sprints.component.tsx [hunk 17]
```diff
@@ -23,6 +23,7 @@ import {
     upcomingSectionId,
 } from '@core/config/sections';
 import { taskUrgencyOrder } from '@core/config/sortOrder';
+import { taskPriorityOrder } from '@core/config/sortOrder';
```
Comment (naturetime10) [inline src/components/pages/Sprints.component.tsx]: Did we use `taskPriorityOrder` in this TypeScript module (a file is a TS module)?

Comment (naturetime10) [inline src/components/pages/Sprints.component.tsx]:
```suggestion
```

**Review by naturetime10 — CHANGES_REQUESTED** [review]

**Review by naturetime10 — APPROVED** [review]

**Pull request merged**